janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java

```diff
@@ -368,14 +384,15 @@ public void mutate(Map<String, Map<String, IndexMutation>> mutations, KeyInforma
 
                         if (isNewDoc)
                             logger.trace("Adding new document {}", docId);
-
-                        for (IndexEntry e : mutation.getAdditions()) {
-                            final Object fieldValue = convertValue(e.value);
-                            doc.setField(e.field, isNewDoc
-                                    ? fieldValue : new HashMap<String, Object>(1) {{
-                                put("set", fieldValue);
-                            }});
-                        }
+                        // Build map of additions
+                        final Map<String,Object> adds = buildFieldAdditions(mutation.getAdditions(), collectionName, informations);
+                        adds.keySet().stream().forEach(v-> {
+                            final KeyInformation keyInformation = informations.get(collectionName, v);
+                            final String key = keyInformation.getCardinality() == Cardinality.SINGLE ? "set" : "add";
```

Comment:
Maybe a different name, like modifier?  'key' can be confused with index key, etc.